Fix iframe.onload being fired twice on about:blank.

gpoesia opened the pull request:
<!-- Please describe your changes on the following line: -->
Supress the load event in the script thread when the iframe doesn't have a src attribute.
In that case, a load event is already fired when binding the iframe to the tree.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15727 (github issue number if applicable).

- [X] There are tests for these changes OR

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16091)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/script_thread.rs
 * @kichjang: components/script/script_thread.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits include an empty title element (`<title></title>`). Consider adding appropriate metadata.

cbrewster commented on tests/wpt/web-platform-tests/html/webappapis/scripting/events/iframe-onload.html:
```diff
@@ -0,0 +1,35 @@
+      }
+    });
+  }, "IFrameSingleOnLoad");
+  document.getElementById('blank_iframe').contentWindow.location.reload();
```
I'm not sure if this causes the same behavior as doing an inline onload on the iframe. I wonder if we could just create and add an iframe via javascript. (I would test this first on the master branch and see if that yields the `onload` event being fired twice).

gpoesia replied:
I had tested it on master and it did trigger two load events. However, I agree it's better to add a new iframe, since it's simpler and doesn't raise doubts. Will do!

cbrewster commented on tests/wpt/web-platform-tests/html/webappapis/scripting/events/iframe-onload.html:
```diff
@@ -0,0 +1,35 @@
+  function iframeOnLoad(t) {
+    iframe_loads++;
+
+    setTimeout(function() {
```
Unfortunately, I am not entirely sure if there is a better option than doing a `setTimeout`.

cbrewster commented on components/script/script_thread.rs:
```diff
@@ -1635,7 +1635,13 @@ impl ScriptThread {
+            Some(iframe) => {
+                // If the iframe has no "src" attribute, a load event will have already been fired
+                // when binding the iframe to the tree.
+                if iframe.upcast::<Element>().has_attribute(&local_name!("src")) {
```
I have been thinking about this and testing a bit more... I don't know if this will be a sufficient solution. Consider the following case:

 * You have the `iframe` with no `src` 
 * The load event steps for the initial `about:blank` are fired async as per the spec
 * In a script you set the `src` of the `iframe` to some URL
 * The load event steps are now processed and a `load` event is fired for the initial `about:blank`

A possible fix is to cancel the load event steps runnable when the src is set.
I have been working on `about:blank` fixes (#14764) and here is a commit that does this https://github.com/servo/servo/pull/14764/commits/46220d409e75ddedefa3585115c3ea89a1e74b25
I would be ok if you wanted to implement something like that here since I don't know how soon I will have enough time to finish up the rest of that PR

gpoesia replied:
Hmm, right, now I see that case. I couldn't reproduce it, though (I tried this:
```html
<html>
  <body>
  </body>
  <iframe id="iframe" onload="alert('onload');"></iframe>
  <script type="text/javascript">
    document.getElementById("iframe").src = "a.html";
  </script>
</html>
```
)

Does the spec say anything on what should happen to the first load event in the case of changing the `src`? I looked at your commit, but if I got it correctly it will only prevent the load steps from being loaded as long as the `IFrameLoadEventSteps` runnable checks the cancel flag after it is set, right? I'm imagining a case in which the program is exactly before `this.iframe_load_event_steps(self.pipeline_id);`, but after the `if` condition has been evaluated, and at this point the src changes and the other thread tries to cancel the event. The check makes it even harder for both events to run, but does it really guarantee it?
If you think it helps I can merge that commit into this PR. What do you think?

cbrewster replied:
So my solution may not be the best way, the spec basically says to delay the load event if a new navigation is triggered: https://html.spec.whatwg.org/multipage/embedded-content.html#iframe-load-event-steps (See the first NOTE).

The code is not running multithreaded here, we just have asynchronous tasks. So when we queue a task to do the iframe load event steps, this will allow other things to potentially be processed first (i.e. src changing before the task is handled). We currently have a `LoadBlocker` for `Document`s that handle this. Maybe we should do something similar with `iframe`s.

cbrewster reviewed (changes requested):
Looks good, although I did find another edge case ☹️

nox commented:
r? @cbrewster

gpoesia commented:
@cbrewster Oh, I see, thanks for the clarifications and patience! For some reason I thought we had multiple threads involved. I've merged your solution into this commit then, and also modified the test to create the iframe in JavaScript instead. Would you take another look?

cbrewster commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -54200,6 +54200,16 @@
      {}
     ]
    ],
+   "html/webappapis/scripting/events/iframe-onload-frame.html": [
```
revert

cbrewster commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -54200,6 +54200,16 @@
+     {}
+    ]
+   ],
+   "html/webappapis/scripting/events/iframe-onload.html2": [
```
revert

cbrewster commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -179815,6 +179831,18 @@
    "491fd73a4ec8812cb8dc1ee01e34a7ff2a07ffb3",
    "testharness"
   ],
+  "html/webappapis/scripting/events/iframe-onload-frame.html": [
```
revert

cbrewster commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -179815,6 +179831,18 @@
+   "d5e1649c88102f27da5fe1ac16715cfee7f70f84",
+   "testharness"
+  ],
+  "html/webappapis/scripting/events/iframe-onload.html2": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -7023,6 +7023,11 @@
      {}
     ]
    ],
+   "css/border_collapse_simple_ref.html2": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -8908,6 +8913,11 @@
      {}
     ]
    ],
+   "css/table_margin_a.html.bak": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -8918,6 +8928,11 @@
      {}
     ]
    ],
+   "css/table_margin_ref.html.bak": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -9583,6 +9598,11 @@
      {}
     ]
    ],
+   "mozilla/referrer-policy/generic/subresource/mozresource.pyc": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -20385,6 +20405,10 @@
    "00c9c85a861dacec7b34f019a43e347c11098ab0",
    "support"
   ],
+  "css/border_collapse_simple_ref.html2": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -23413,6 +23437,10 @@
    "0617c8de92ec1a283ac429dbbb073805e6bad28c",
    "reftest"
   ],
+  "css/table_margin_a.html.bak": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -23425,6 +23453,10 @@
    "d2ccfd91d67914e7e6653e1ce86deeebf5cd5977",
    "support"
   ],
+  "css/table_margin_ref.html.bak": [
```
revert

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -25737,6 +25769,10 @@
    "333a80e3d4aec4a8be5566605fae5675a821085b",
    "support"
   ],
+  "mozilla/referrer-policy/generic/subresource/mozresource.pyc": [
```
revert

cbrewster commented on tests/wpt/web-platform-tests/html/webappapis/scripting/events/iframe-onload.html:
```diff
@@ -0,0 +1,37 @@
+      function iframeOnLoad(t) {
+        iframe_loads++;
+
+        setTimeout(function() {
```
does this test fail on master? I think 
```
t.step(function() {
    assert_true(iframe_loads === 1, "iframe.onload should be called exactly once.");
});
```
should be above the timeout

cbrewster reviewed (changes requested):
Looks like there were some unintended changes in the `MANIFEST`

cbrewster commented:
I have been thinking about this more, there are still edge cases where using the `cancelled` flag won't work. We may need to develop a more sophisticated way to `delay` iframe load events. Possibly something similar to document's `LoadBlocker` but a more simplified version.

cbrewster commented:
@gpoesia sorry about the back and forth on this one... I believe you can remove my changes and this should be good to go. I forgot that when firing a load event, we check to see if the pipeline of the child document that was loaded matches the latest pipeline of the iframe. This already handles the case I mentioned above.

gpoesia commented:
@cbrewster no problem! I've fixed the unwanted changes in the MANIFEST files, and reverted the cancel flag part as well. 

About the test, I've checked again and it does fail on master. The idea was waiting to call `t.step()` to give some time for the next load event to run, should it be fired. Then the assertion fails on both calls to `t.step()` (one from each event). If I put the assertion outside the timeout, then the first event will set a timeout to call `t.done()` directly, while the other would fail. Then it depends on what is stronger - the failed assertion or the delayed call to `t.done()`. Do you prefer one over the other? Let me know and I can change it as well.

cbrewster commented:
@gpoesia looks good! Thanks!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 09d2fad has been approved by `cbrewster`

<!-- @bors-servo r=cbrewster 09d2fadb70ed6edc00beef81ef0c9fcb9a822b22 -->

bors-servo commented:
:hourglass: Testing commit 09d2fadb70ed6edc00beef81ef0c9fcb9a822b22 with merge c7b39d08944b0233c7c04c96847d5c97b5e70e6b...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/6043)

jdm commented:
`./mach update-manifest` is required.

gpoesia commented:
Did it and forgot to comment here. It seems at least the CI checks now pass.

cbrewster commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 3f69bc3 has been approved by `cbrewster`

<!-- @bors-servo r=cbrewster 3f69bc3c88616db783df7122667a9b0dc5fe8211 -->

bors-servo commented:
:hourglass: Testing commit 3f69bc3c88616db783df7122667a9b0dc5fe8211 with merge dea905c4fe181b363ac753a6988ebd7d90c8b9c4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3166)

jdm commented:
The html5lib tests were _really_ unhappy about these changes. /html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent.html may be an easier one to start investigating.

cbrewster commented:
My guess is that now some `onload` events are not firing when they used to which causes these tests to `NOTRUN`. I haven't had the time to figure the exact scenario with the html5lib tests. It could be that these tests were only passing because `onload` was fired twice unintentionally.

gpoesia commented:
It seems one problem (at least with `navigate-child-function-parent.html`) is that there are other ways of specifying an URL to be loaded other than by setting a `src`. This test specifically sets `iframe.contentWindow.location`, but then the load event steps are supressed because the iframe doesn't have an 'src' attribute, which is checked for. Hence, I think we should also check if `contentWindow.location` is unset before deciding not to fire the event. I've added that to the PR. It seems at least that test has the expected result now.

cbrewster commented:
@bors-servo try
I'm not sure if we should special case so much here.

bors-servo commented:
:hourglass: Trying commit 047c9eadfb331e7532fe8735cf448913357801ed with merge bdc70eabca59fed366ba42adc54bc052da159bf2...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1805)

cbrewster commented:
```
error[E0277]: the trait bound `&url::Url: core::cmp::PartialEq<core::option::Option<_>>` is not satisfied
    --> C:\buildbot\slave\windows-msvc-dev\build\components\script\script_thread.rs:1667:66
     |
1667 |                 if element.has_attribute(&local_name!("src")) || window.get_url().as_url() != None {
     |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `core::cmp::PartialEq<core::option::Option<_>>` is not implemented for `&url::Url`
     |
     = help: the following implementations were found:
               <url::Url as core::cmp::PartialEq>
```

gpoesia commented:
Weird, I don't get that locally. I've changed it to use pattern matching instead.

gpoesia commented:
Hold on a sec, I'll need one more change before we can try again.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16528) made this pull request unmergeable. Please resolve the merge conflicts.

gpoesia commented:
@cbrewster I've also added a `DocumentType` (borrowing the idea from your previous PR), which reaches the ScriptThread and is used for deciding whether to fire the load event or not. Should I wait to rebase it on top of #16506?

cbrewster commented:
@gpoesia that would be great! I would wait until it at least gets r+ed or lands since I still need to make a few adjustments.

cbrewster commented:
@gpoesia #16506 has landed now. Are you still interested in working on this?

gpoesia commented:
@cbrewster Sure thing! I will come back to this this week.

gpoesia commented:
@cbrewster I'm back on this. I've pushed another solution, based on tracking the first load event and suppressing it when parsing finishes. I've tested it with and without an src and it seems to work.

`/wpt/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent.html` still fails, though. However, I think there's more to it than this PR. It seems you were right when you said this test might only be passing because two load events were being incorrectly fired. Setting `iframe.contentWindow.location` is *not* causing the inner document to load. I've tested it with a simple setup:

h1.html:
```html
<h1>Test</h1>
```

contentWindow.location.html:
```html
<!doctype html>
<meta charset=utf-8>
<title>Set location from a parent</title>
<iframe></iframe>
<script>
  onload = function() {
    var fr = document.querySelector("iframe")
    fr.contentWindow.location = "h1.html"
    fr.onload = function() {
       console.log("Loaded");
    }
  }
</script>
```

This was derived from `navigate-child-function-parent.html`. Opening it either in master or my branch doesn't actually load `h1.html` (in Firefox it does).

Do you agree that is for another PR? Otherwise, I can start looking into it.

cbrewster commented:
@gpoesia thanks for picking this back up! I would expect `/wpt/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function-parent.html` to fail as we do not implement `PutForwards` for `Window` in the web IDL see https://dxr.mozilla.org/servo/rev/c17134ce485453ff124cc337fa68b9a981106a45/components/script/dom/webidls/Window.webidl#16

The `PutForwards` attribute is what allows `Location` to have a special behavior when assigned to. See https://heycam.github.io/webidl/#PutForwards.

I do not know what would need to be done to implement `PutForwards` at this time, but it shouldn't be done as a part of this PR.

gpoesia commented:
All right, got it! Should I change the test expectations then? I remember there were others as well, but the link to the builder log has expired. Can we @bors-servo try once more so that we have the list of failing tests again? I'll then see if they are all related to `contentWindow.location` or if it's something else I missed.

bors-servo commented:
@gpoesia: :key: Insufficient privileges: and not in try users

gpoesia commented:
@bors-servo it wasn't meant to be an order, sorry if it sounded like one!

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 317e9e33dbc39ac8fecbe6a651d2f161f5f19648 with merge 72d5aeb66ea406a8d82842abff5eafe93839b008...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/596)

gpoesia commented:
I'm not sure what to make of `/tests/wpt/mozilla/tests/css/layerization_layer_size.html`. It does have an empty iframe, but no JavaScript involved. I can render it normally using `./mach run`. Any ideas on how to start debugging a timeout like this one? The test does indeed pass on master locally. Does the test harness use the onload event of the iframe implicitly?

jdm commented:
Yes; for reference tests the test harness takes a screenshot of the rendered page once the window's load event has been fired, and then does the same for the reference file mentioned in the `<link rel="match">`.

jdm commented:
And since the window's load event will not be fired until the iframe's load event has been fired, a timeout likely means that the iframe's load event did not fire.

gpoesia commented:
Makes sense. All right, I'll start looking. Thanks, @jdm!

gpoesia commented:
OK, at least that test is fixed now. I tried suppressing the event earlier, but it turns out the second reflow triggered by `iframe_load_event_steps` was important for not running into #13149. Otherwise, WebRender would never become aware of the new pipeline ID, and the compositor would never paint since the Constellation would see that the Compositor is not tracking the iframe's pipeline (in `Constellation::handle_is_ready_to_save_image`).

If someone can think of a simpler solution, I'll be happy to implement it.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17735) made this pull request unmergeable. Please resolve the merge conflicts.

gpoesia commented:
Rebased. Can we try again?

cbrewster commented:
@bors-servo try
Sorry for the delay! I will review this tomorrow.

bors-servo commented:
:hourglass: Trying commit bd87df6b799e774dc5b59d88082fb3680551ca5f with merge 1534dbc18067ef968f2ca8a1655c6ef183ee32ae...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5110)

paulrouget commented:
I haven't really looked a this code and PR, I'm just curious: Is it necessary to propagate the information that it is the initial about:blank page that has loaded?

Could there be a way, in `handle_subframe_loaded`, from the pipeline and browsing context, to know that it is the initial about:blank page? If not, what about adding a flag to the browsing context or the pipeline when it's created?

cbrewster commented:
@paulrouget's idea may be the best way to do this. Ideally, we could stay as close to the spec as possible, but AFAICT we don't follow https://html.spec.whatwg.org/multipage/#creating-a-new-browsing-context exactly (someone who knows more about this could correct me). Trying to propagate the information about whether the document/load is the initial about:blank load in the script thread is a bit messy. It might be cleaner to keep track of this constellation and abort `handle_subframe_loaded` if the load the completed was the initial about:blank load.

I am interested what @asajeffrey thinks about this issue, IIRC he was not fond of tracking the initial about:blank info in the constellation if it can be avoided.

gpoesia commented:
Thanks, @paulrouget and @cbrewster. I also prefer a simpler solution if we can find one. The issue I ran into when trying to suppress the load event earlier (e.g. in `handle_subframe_loaded`) was the behavior described in #13149. Apparently, it was only when the second load event was processed that WebRender would become aware of the new pipeline ID. If I suppress the event earlier, what happens is that while the Constellation tracks new pipeline (i.e. it is listed in `Constellation::fully_active_browsing_contexts_iter`), the Compositor does not. As a result, `Constellation::handle_is_ready_to_save_image` always falls into the `ReadyToSave::PipelineUnknown` case, which causes the Compositor to never paint. Since it never paints, WPT tests with iframes time out (they just hang with the Compositor waiting for Constellation's signal that it's ready to paint).

Actually, it's even a bit more subtle. The compositor itself is notified of the new pipeline and tracks it (in `Compositor::IOCompositor<>::pipeline_details`. However, before sending the "ReadyToSaveImage" query to the constellation, it matches the pipeline ids it has with the ones WebRender has epochs for. This is what is false in the blank iframe case: WebRender has no epoch for the new pipeline ID, which causes the compositor not to send the pipeline ID to the constellation along with the query, for which constellation returns `ReadyToSave::PipelineUnknown`.

The way I found around this was to let the event be processed almost until the very end, and only suppress the part in which the "load" event is fired on the element. Right after that, a reflow happens and I could not make it work without letting it run (the second time `iframe_load_event_steps` is called, from ScriptThread).

But we can probably stop earlier, as long as we work around this issue with the compositor somehow. Then, I think stopping in `handle_subframe_loaded` should work. @paulrouget @cbrewster, any ideas on how to do that?

I didn't track the problem much further than that. I can debug the reflow and see what is in it that causes the problem not to happen, so that we may do just that when suppressing the event.

asajeffrey commented:
@cbrewster indeed, I was happier about keeping the constellation out of the whole initial about:blank situation if possible! We need to special-case the initial about:blank anyway, to make it synchronous, so I was hoping we could hide the special-casing from the constellation. But this may not be feasible.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17634) made this pull request unmergeable. Please resolve the merge conflicts.

gpoesia commented:
Rebased and fixed the crash. I'm up for discussing alternatives, of course.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17425) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@cbrewster Do you have time to continue reviewing this?

cbrewster commented:
I do, but I'm not entirely sure what the correct, spec-compliant approach is for fixing this issue. It would be good to have someone else look into it.

jdm commented:
r? @asajeffrey 
Alan, could you take this over?

asajeffrey commented:
Back from vacation, I'll have a look over this.

asajeffrey commented:
Some IRC chat with @jdm: http://logs.glob.uno/?c=mozilla%23servo&s=7+Sep+2017&e=7+Sep+2017#c746388

TL;DR we shouldn't be navigating to the initial about:blank, we should just have it magically appear. The problem line is https://github.com/servo/servo/blob/1e93749941d2e3569c7c907832658c57ffb18c72/components/script/dom/htmliframeelement.rs#L269

asajeffrey commented:
Hmm, we're already navigating the newly created browsing context when we process the iframe attributes: https://github.com/servo/servo/blob/e7f45028dcd4d7015b4eed2ecc193e176b1fbd1c/components/script/dom/htmliframeelement.rs#L253

Can we remove the navigation when we create the browsing context? https://github.com/servo/servo/blob/e7f45028dcd4d7015b4eed2ecc193e176b1fbd1c/components/script/dom/htmliframeelement.rs#L284

If we remove that navigation, there may be some extra code needed to handle the first-load-with-no-src case https://github.com/servo/servo/blob/e7f45028dcd4d7015b4eed2ecc193e176b1fbd1c/components/script/dom/htmliframeelement.rs#L235-L238

asajeffrey commented:
Doing a bit more digging...

If we added a field to `InProgressLoad` tracking whether or not the document is an initial `about:blank`, then this would be available to `script_thread::load`, which creates the `Document` object. This might be the right place to suppress the second load event without needing any changes to the constellation.

asajeffrey commented:
And some more digging... if we add another option to `DocumentSource` for initial `about:blank` documents, and store the document source in the `Document`, then we can change `DocumentProgressHandler::dispatch_load` to be something like:

```rust
     fn dispatch_load(&self) {
         let document = self.addr.root();
         if document.browsing_context().is_none() || document.source == DocumentSource::InitialAboutBlank {
             return;
         }
         ...
```

This way the constellation won't get asked to forward the load event on to the parent.

@gpoesia I realize this is a very different approach from the one you were using before, does this sound sensible to you?

asajeffrey commented:
@gpoesia thoughts?

asajeffrey commented:
@gpoesia ping?

gpoesia commented:
Thanks for looking into it, @asajeffrey!
I remember the issue I ran into when trying to stop the load event earlier was this one I described here: https://github.com/servo/servo/pull/16091#issuecomment-316697008

Basically, the reflow that was caused by the second load event was actually important. Without it, the compositor thread wouldn't ever paint the iframe (and tests would hang, as a consequence). I didn't figure out why that was the case, and it's even possible it isn't true anymore.

I really like the solution of simply adding another `DocumentSource`. I'll have some time to implement it today and we can see if we run into same issue. Even if we do, we can try to figure out a way to circumvent it without polluting the whole code base with a special case for initial about:blanks. I'll be able to do that this weekend if need be.

gpoesia commented:
@asajeffrey this is what would happen when the load event was suppressed before the second reflow: https://github.com/servo/servo/issues/13149

asajeffrey commented:
Hmm, so we need to find a way to trigger a reflow without a second load event.

BTW, heads up that #18587 might make a difference here!

gpoesia committed:
Suppress second iframe about:blank load event in Document.

gpoesia commented:
@asajeffrey I've updated the code, is that more or less what you had in mind?

Unfortunately, the `navigate-child-function-parent.html` WPT test also times out with this patch. I thought there was a chance that the reflow that happens in `Document::maybe_queue_document_completion` would suffice, if it was only that a second reflow was needed. However, I believe it's more subtle than that.

From what I could gather, we need a reflow that happens after the compositor has been notified of the iframe's existence and latest pipeline ID. Since the data flow is always `Module <--> Constellation`, the only way of the compositor knowing is via constellation (I think the crucial part is the message it receives from `Constellation::handle_load_complete`). I think a reflow is necessary after that message. Incidentally, one would happen in the second time `HTMLIFrameElement::iframe_load_event_steps` is called, which takes place down the rabbit role in `Constellation::handle_load_complete`, so tests would pass before.

I guess receiving a `LoadComplete` message is all the Compositor needs, but I'm not sure. My last attempt was on the safer side of letting everything that would happen in the second load event take place, except for the actual firing of the event in the element.

So:

1- Does my code implement now what you had in mind? If it doesn't, I can fix it and try to make it work.
2- If it does and you also believe there's no way of not letting the Constellation know the second load event happened, do you see another place where it would be natural to trigger a reflow before the event reaches the iframe element again?

Another possible solution I can try is to simply store a flag in `HTMLIFrameElement`, which says whether further load events should be suppressed. We can update this flag in `HTMLIFrameElement::process_the_iframe_attributes`, setting it to true in the initial `about:blank` special case and false otherwise. Then, we check for it in `iframe_load_event_steps` and do the same I was doing before, of letting the reflow happen regardless of whether the event was fired. I'll see if I can make that work today. It was one of the first things I thought when looking at the issue, but by that time it clearly seemed to be an ugly hack that didn't go to the problem's root. But if it turns out to be the case that everything in the path to the second load event except the firing itself is needed, it might just be simpler. And now I can at least explain the whole reason behind the flag in a comment. Does it sound reasonable?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18635) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
Yes, this is the sort of thing I had in mind, yay!

There are a few nits, like it would be nice to use an enum rather than a bool for whether a document is the initial about:blank, but we can sort those out once we've worked out how to get the tests to pass.

It sounds like you have a point, we need a way for the constellation to know that the iframe has finished loading. Is this just for `constellation::handle_is_ready_to_save_image`, or do we need a reflow for correct execution in non-test environments?

One thing which has occurred to me is that `constellation::handle_script_new_iframe` shouldn't add a pending change. This would also have the effect of allowing save image earlier, but I'm concerned that it'll end up being too early, or we'll end up with weird race conditions.

gpoesia commented:
As far as I could tell, this is all to satisfy `constellation::handle_is_ready_to_save_image`, yes. It might be the case that it serves other purposes as well, but the issue I ran into specifically happened because `constellation::handle_is_ready_to_save_image` would always return `PipelineUnknown`. But we can try to fix just that and see whether it breaks somewhere else :)

I see two ways of going about the Constellation issue: we can create a "special path" for the iframe load to reach the constellation in that case, or we can just extend the path with more flags/enums so that the event stops earlier than usual. I think this latter one is less special-cased than creating a new path (e.g. a new message type for the constellation). I'm not sure about where the reflow would fit, though. Am I right in thinking that only the script thread has enough information to issue a reflow?

@asajeffrey Do you think adding the flag in `HTMLIFrameElement` is too hacky? :)
I've implemented it and it seems to work.

asajeffrey commented:
@gpoesia which flag on `HTMLIFrameElement` were you thinking of? If it's one for "does the iframe contain the initial about:blank" can we do this by finding the nested document and asking it if it's the initial about:blank? I'd rather do that than keep track of more mutable state.

asajeffrey commented:
@gpoesia what would be the best way to move forward on this?

asajeffrey commented:
@gpoesia Are you still working on this?

asajeffrey commented:
This looks dormant, feel free to reopen it!
